src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+// This class observes those counters and keeps a snapshot of readings. For now this is used only to print cache stats in debug mode.
+// TODO: We could add some System.Diagnostics.Metrics.Gauge instruments to this class, to get computed stats also exposed as metrics.
+type CacheMetrics(cacheId) =
+    static let meter = new Meter("FSharp.Compiler.Cache")
+
+    static let observedCaches = ConcurrentDictionary<string, CacheMetrics>()
+
+    let readings = ConcurrentDictionary<string, int64 ref>()
+
+    let listener = new MeterListener()
+
+    do
+        listener.InstrumentPublished <-
+            fun i l ->
+                if i.Meter = meter && i.Description = cacheId then
+                    l.EnableMeasurementEvents(i)
+
+        listener.SetMeasurementEventCallback<int64>(fun k v _ _ -> Interlocked.Add(readings.GetOrAdd(k.Name, ref 0L), v) |> ignore)
+        listener.Start()
+
+    member this.Dispose() = listener.Dispose()
+
+    member val CacheId = cacheId
+
+    static member val Meter = meter
+
+    member val RecentStats = "-" with get, set
+
+    member this.TryUpdateStats(clearCounts) =
+        let ratio =
+            try
+                float readings["hits"].Value
+                / float (readings["hits"].Value + readings["misses"].Value)
+                * 100.0
+            with _ ->
+                Double.NaN
+
+        let stats =
+            [
+                for name in readings.Keys do
+                    let v = readings[name].Value
+
+                    if v > 0 then
+                        $"{name}: {v}"
+            ]
+            |> String.concat ", "
+            |> sprintf "%s | hit ratio: %s %s" this.CacheId (if Double.IsNaN(ratio) then "-" else $"%.1f{ratio}%%")
+
+        if clearCounts then
+            for r in readings.Values do
+                Interlocked.Exchange(r, 0L) |> ignore
+
+        if stats <> this.RecentStats then
+            this.RecentStats <- stats
+            true
+        else
+            false
+
+    // TODO: Should return a Map, not a string
+    static member GetStats(cacheId) =
+        observedCaches[cacheId].TryUpdateStats(false) |> ignore
+        observedCaches[cacheId].RecentStats
+
+    static member GetStatsUpdateForAllCaches(clearCounts) =
+        [
+            for i in observedCaches.Values do
+                if i.TryUpdateStats(clearCounts) then
+                    i.RecentStats
+        ]
+        |> String.concat "\n"
+
+    static member AddInstrumentation(cacheId) =
+        if observedCaches.ContainsKey cacheId then
+            invalidArg "cacheId" $"cache with name {cacheId} already exists"
+
+        observedCaches[cacheId] <- new CacheMetrics(cacheId)
+
+    static member RemoveInstrumentation(cacheId) =
+        observedCaches[cacheId].Dispose()
+        observedCaches.TryRemove(cacheId) |> ignore
+
+type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
```

Comment:
What is the starting size?

Reply:
Zero, we just allocate items as requested up to totalCapacity. Once they're removed from the cache we put them into the bag. If there was a resize because of burst activity and we had to allocate more, we still throw away the excess, and let it be collected. In effect the bag holds only up to totalCapacity at any moment.